APPS-303 + APPS-351 auto-detached subjobs + dxfuse update

mhrvol opened the pull request:
Subjobs in cwic are started as detached if $cmd is not passed as argument to cwic (more precisely, DX_RUN_DETACH is set to $INTERACTIVE from cwic).

This means that if used starts cwic with $cmd AND with --ssh flag, subjobs will not run as detached. Is this correct?
@emiloslavsky 

Also note that DX_RUN_DETACH is set only in docker, not on the worker - if user exits from docker, jobs will not be detached

commandlinegirl commented:
Setting it only within Docker env should be sufficient, since this is the target environment for cwic.

I think the variable will also need to be set every time the user loads the environment (?). E.g. when they open a byoby terminal (see my comment in file-apps for the cloud workstation)?

mhrvol commented:
> I think the variable will also need to be set every time the user loads the environment (?). E.g. when they open a byoby terminal (see my comment in file-apps for the cloud workstation)?

TCs:
dx run cwic --ssh > dx_run_detach is set
dx ssh <job>.  > dx_run_detach is set
new window in terminal > dx_run_detach is set

emiloslavsky commented:
@commandlinegirl please re-review

commandlinegirl reviewed (approved):
LGTM

commandlinegirl commented:
Actually, two additional things need to be done for this to complete:

1. we need to upgrade dxpy in the [Docker image](https://github.com/dnanexus/cwic/blob/main/docker/Dockerfile#L38). Please pin the version so we know in the future what we installed. 

2. in [dx-cwic-sub](https://github.com/dnanexus/cwic/blob/main/resources/usr/local/bin/dx-cwic-sub#L38-L52) - we currently require a dx token to start it. With detached jobs this is not needed. How about we check the dx version, if it's lower than one that contains the auto-detached job, we print an error as we do today (with the error messages extended with a note to upgrade dxpy to be able to run in autodetached mode). If the version is higher than xxx, then we skip this whole thing and proceed. cc @emiloslavsky

commandlinegirl commented:
> Subjobs in cwic are started as detached if $cmd is not passed as argument to cwic (more precisely, DX_RUN_DETACH is set to $INTERACTIVE from cwic).
> 
> This means that if used starts cwic with $cmd AND with --ssh flag, subjobs will not run as detached. Is this correct?
> @emiloslavsky

@mhrvol We discussed it today and cwic will be running with DX_DETACHED_JOB=1 regardless whether it's interactive or non-interactive. Please see also my other comment re dx-cwic-sub.

> Also note that DX_RUN_DETACH is set only in docker, not on the worker - if user exits from docker, jobs will not be detached

That's fine. cwic is meant to be used inside the Docker environment. Thanks!

mhrvol commented:
> 1. we need to upgrade dxpy in the [Docker image](https://github.com/dnanexus/cwic/blob/main/docker/Dockerfile#L38). Please pin the version so we know in the future what we installed.

I've pinned the version, updated the docker image and updated the default docker image in cwic

> 2. in [dx-cwic-sub](https://github.com/dnanexus/cwic/blob/main/resources/usr/local/bin/dx-cwic-sub#L38-L52) - we currently require a dx token to start it. With detached jobs this is not needed. How about we check the dx version, if it's lower than one that contains the auto-detached job, we print an error as we do today (with the error messages extended with a note to upgrade dxpy to be able to run in autodetached mode). If the version is higher than xxx, then we skip this whole thing and proceed. cc @emiloslavsky

DNAnexus token is now requested only in case user wants the job to be run as a detached job AND detached jobs are supported by current dxpy. 

If current dxpy does not support detached jobs, there is a message about upgrading dx-toolkit and jobs will not run in detached mode - as there is not support for that (but there is not hard-exit as I thought it might cause some changes in current scrips of our users.) 

Let me know if this is acceptable/correct

@commandlinegirl

mhrvol commented:
> We discussed it today and cwic will be running with DX_DETACHED_JOB=1 regardless whether it's interactive or non-interactive. Please see also my other comment re dx-cwic-sub.

DX_RUN_DETACH is always set to 1 now

commandlinegirl commented on resources/usr/local/bin/dx-cwic-sub:
```diff
@@ -35,20 +42,27 @@ if [[ "$LAST_ARG" == --* ]]; then
-        echo -e "DNAnexus token is not set in the credentials file. Please provide DNAnexus token or log in. Exiting..."
-        exit 1
+if [ $ABLE_TO_RUN_DETACH = 0 ]; then
+    echo -e "Please update the version of dx to be able to run jobs in detached mode. To update dx use 'pip install dxpy --upgrade'"
```
.. the version of dxpy in your Docker image to be able ..

mhrvol replied:
fixed

commandlinegirl commented on dxapp.json:
```diff
@@ -59,7 +59,7 @@
   },
   "details": {
-    "whatsNew": "* 0.0.6: Add dx-find-cwic-jobs and dx-cwic-sub scripts; mount project in read only mode by default; set HOME to /home/cwic; upgrade dxfuse to v0.22.2\n* 0.0.5: Upgrade dxfuse to v0.22.1\n* 0.0.4: Load cwic when opening a tmux window; Use dnanexus/cwic-base:0.0.1 as default image; Fix prompt; Bug fixes\n* 0.0.3: Bug fixes\n* 0.0.2: Pass DNAnexus user token in credentials and use i ...
+    "whatsNew": "* 0.1.0: Jobs started from Cloud workstation are started in detached mode by default. Updated version of dx-toolkit to v0.305.0. \n* 0.0.6: Add dx-find-cwic-jobs and dx-cwic-sub scripts; mount project in read only mode by default; set HOME to /home/cwic; upgrade dxfuse to v0.22.2\n* 0.0.5: Upgrade dxfuse to v0.22.1\n* 0.0.4: Load cwic when opening a tmux window; Use dnanexus/cwic ...
```
If you can also and Updated base image to 0.0.4 (and also bump up the version here: https://github.com/dnanexus/cwic/blob/main/docker-compose.yml#L8) and run build.sh.

mhrvol replied:
build.sh has "dnanexus/cwic-base:latest" in it. that should stay that way, right?

commandlinegirl replied:
That's good. docker-compose push already pushed the numbered version and it's ok to already push before app release in this case.

commandlinegirl commented on resources/usr/local/bin/dx-cwic-sub:
```diff
@@ -13,6 +13,13 @@ JOB_INPUT_FILE="${DNANEXUS_HOME}/job_input.json"
 LAST_ARG=${*: -1}
+DX_VERSION=$(dx --version | sed 's/.* //')
+FIRST_DETACHED_VERSION="v0.305.0"
+if [[ ($FIRST_DETACHED_VERSION != $DX_VERSION && $FIRST_DETACHED_VERSION = "`echo -e "$FIRST_DETACHED_VERSION\n$DX_VERSION" | sort -V | head -n1`") ]]; then
```
if DX_VERSION == "v0.305.0", ABLE_TO_RUN_DETACH will be set 0.

mhrvol replied:
Tested it in a different test file and mixed up 1 and 0 :). fixed

commandlinegirl replied:
np, but I think now it'll fail when DX_VERSION > "v0.305.0"

mhrvol replied:
Also in the test file I've been changing wrong value for testing :) 

Thanks for noticing!

commandlinegirl reviewed (approved):
LGTM. Please upgrade cwic and publish, thank you!

mhrvol commented:
I did a little bit more extensive testing today, found issues with dx-cwic-sub if we do not login beforehand.

if applet is created and dx-cwic-run is run : 
```dxpy.utils.resolver.ResolutionError: No accessible project could be found containing the entity applet-G00gvg00V3b41q5k2X6p7kPZ, code 401. Request Time=1610710812.4009132, Request ID=1610710812417-926752```
	If app is created and dx-cwic-run is run: 
```dxpy.utils.resolver.ResolutionError: No matches found for app-G00g5jj9pfqKYY2ZKqKFK2K9```

dx-cwic-sub and/or cwic interactive usage without credentials:
	if published CWIC is started: 
```dxpy.utils.resolver.ResolutionError: At least VIEW permission is required to search within a project, code 401. Request Time=1610711330.9052026, Request ID=1610711330914-623041```


I suggest we return to the old way of dealing with credentials in dx-cwic-sub and just start the job in detached mode if possible - we still require dnannexus token for starting CWIC jobs - other jobs work just fine (jobs that do no require certain access in dxapp.json)

commandlinegirl commented:
> I did a little bit more extensive testing today, found issues with dx-cwic-sub if we do not login beforehand.
> 
> if applet is created and dx-cwic-run is run :
> `dxpy.utils.resolver.ResolutionError: No accessible project could be found containing the entity applet-G00gvg00V3b41q5k2X6p7kPZ, code 401. Request Time=1610710812.4009132, Request ID=1610710812417-926752`
> If app is created and dx-cwic-run is run:
> `dxpy.utils.resolver.ResolutionError: No matches found for app-G00g5jj9pfqKYY2ZKqKFK2K9`
> 
This is expected. The job with its token doesn't have access to the cwic built as an applet or an unpublished app. 

> dx-cwic-sub and/or cwic interactive usage without credentials:
> if published CWIC is started:
> `dxpy.utils.resolver.ResolutionError: At least VIEW permission is required to search within a project, code 401. Request Time=1610711330.9052026, Request ID=1610711330914-623041`
> 
When does this error happen? The app has contribute access to the parent project, hmm. Which project might be searched here? It might also be because the app is not published. I suggest we publish the app (with the changes to dxfuse below) and test. If things break, we can release another version.

> I suggest we return to the old way of dealing with credentials in dx-cwic-sub and just start the job in detached mode if possible - we still require dnannexus token for starting CWIC jobs - other jobs work just fine (jobs that do no require certain access in dxapp.json)

Re dxfuse:

With the newest version of dxfuse mounting is done in readOnly mode by default. There is a new flag `-writeOnly` available https://github.com/dnanexus/dxfuse/blob/28c85d0b5f958a49005de8159668c515200a7059/cli/main.go#L58 so this will need to be added to the choices in
https://github.com/dnanexus/cwic/blob/main/dxapp.json#L43 (we can keep -w for backwards comp.). 

BTW, the newest dxfuse, released 2 hrs ago is 0.23.2. If you'd like you can separate these changes into a different PR; if not, please add both JIRA ticket IDs to the commit message. Thx!

mhrvol commented:
> Re dxfuse:
> 
> With the newest version of dxfuse mounting is done in readOnly mode by default. There is a new flag `-writeOnly` available https://github.com/dnanexus/dxfuse/blob/28c85d0b5f958a49005de8159668c515200a7059/cli/main.go#L58 so this will need to be added to the choices in
> https://github.com/dnanexus/cwic/blob/main/dxapp.json#L43 (we can keep -w for backwards comp.).

I cannot see "writeOnly", only "readWrite" which is basically -w, isn't it? I've added "-readWrite" as an option in cwic, but it behaves the same way as -w

mhrvol commented:
> > dx-cwic-sub and/or cwic interactive usage without credentials:
> > if published CWIC is started:
> > `dxpy.utils.resolver.ResolutionError: At least VIEW permission is required to search within a project, code 401. Request Time=1610711330.9052026, Request ID=1610711330914-623041`
> 
> When does this error happen? The app has contribute access to the parent project, hmm. Which project might be searched here? It might also be because the app is not published. I suggest we publish the app (with the changes to dxfuse below) and test. If things break, we can release another version.

This error happens when I run code from this branch, but comment out resolving correct app-id (just force running app-cwic in dx-cwic-sub)

commandlinegirl commented on dxapp.json:
```diff
@@ -59,7 +60,7 @@
   },
   "details": {
-    "whatsNew": "* 0.0.6: Add dx-find-cwic-jobs and dx-cwic-sub scripts; mount project in read only mode by default; set HOME to /home/cwic; upgrade dxfuse to v0.22.2\n* 0.0.5: Upgrade dxfuse to v0.22.1\n* 0.0.4: Load cwic when opening a tmux window; Use dnanexus/cwic-base:0.0.1 as default image; Fix prompt; Bug fixes\n* 0.0.3: Bug fixes\n* 0.0.2: Pass DNAnexus user token in credentials and use i ...
+    "whatsNew": "* 0.1.0: Jobs started from Cloud workstation are started in detached mode by default. Updated version of dx-toolkit to v0.305.0. Upgraded dxfuse to v0.23.1. Updated base image to 0.0.4. \n* 0.0.6: Add dx-find-cwic-jobs and dx-cwic-sub scripts; mount project in read only mode by default; set HOME to /home/cwic; upgrade dxfuse to v0.22.2\n* 0.0.5: Upgrade dxfuse to v0.22.1\n* 0.0.4 ...
```
in cloud ws --> in cwic

mhrvol replied:
fixzed

commandlinegirl commented on dxapp.json:
```diff
@@ -36,11 +36,12 @@
       "class": "string",
       "optional": true,
-      "help": "Mount options for project, set to '-w' in order to mount project with write access (experimental) otherwise read only mode will be used",
+      "help": "Mount options for project, set to '-w' or '-readWrite' in order to mount project with write access (experimental) otherwise read only mode will be used",
```
But the code in this script also needs to change: https://github.com/dnanexus/cwic/blob/main/resources/usr/local/bin/dxfuse-mount#L8

dxfuse mounts in readonly by default now. To mount in rw, you need to set the -readWrite flag. Please check dxfuse docs/code if needed.

mhrvol replied:
Added, tested on -readWrite flag, I was able to create directories and files (after dxfuse-sync)

commandlinegirl commented on resources/usr/local/bin/dx-cwic-sub:
```diff
@@ -13,6 +13,13 @@ JOB_INPUT_FILE="${DNANEXUS_HOME}/job_input.json"
+    ABLE_TO_RUN_DETACH=0
+else
+    ABLE_TO_RUN_DETACH=1
+fi
```
* Help in this script needs update, too..
* If credentials are provided, use them to run the app(let). If not, run the app in detached more (if dx version allows), if not exit as we doing it now) - as we discussed.

mhrvol replied:
Updated. Note that for running in detached mode without dnanexus token/ login, we need to source environment.

commandlinegirl reviewed (approved):
LGTM
